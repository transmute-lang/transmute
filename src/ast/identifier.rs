use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Identifier {
    name: String,
    span: Span,
}

impl Identifier {
    pub fn new(name: String, span: Span) -> Self {
        Self { name, span }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}
