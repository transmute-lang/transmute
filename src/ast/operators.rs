use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct BinaryOperator {
    kind: BinaryOperatorKind,
    location: Location,
    span: Span,
}

impl BinaryOperator {
    pub fn new(kind: BinaryOperatorKind, location: Location, span: Span) -> Self {
        Self {
            kind,
            location,
            span,
        }
    }

    pub fn kind(&self) -> &BinaryOperatorKind {
        &self.kind
    }
}

#[derive(Debug, PartialEq)]
pub enum BinaryOperatorKind {
    Addition,
}
