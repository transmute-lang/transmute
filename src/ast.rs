pub mod expression;
pub mod identifier;
pub mod literal;
pub mod operators;
pub mod statement;

use crate::ast::expression::{ExprId, Expression, ExpressionKind};
use crate::ast::identifier::IdentId;
use crate::ast::literal::{Literal, LiteralKind};
use crate::ast::statement::{Statement, StatementKind, StmtId};
use std::fmt::{Display, Formatter};

pub trait Visitor<R> {
    fn visit_ast(&mut self, ast: &Ast) -> R;

    fn visit_statement(&mut self, stmt: StmtId) -> R;

    fn visit_expression(&mut self, expr: ExprId) -> R;

    fn visit_literal(&mut self, literal: &Literal) -> R;
}

#[derive(Debug, PartialEq)]
pub struct Ast {
    identifiers: Vec<String>,
    expressions: Vec<Expression>,
    statements: Vec<Statement>,
    // todo replace with Statements
    root: Vec<StmtId>,
}

impl Ast {
    pub fn new(
        identifiers: Vec<String>,
        expressions: Vec<Expression>,
        statements: Vec<Statement>,
        root: Vec<StmtId>,
    ) -> Self {
        assert!(!root.is_empty());
        Self {
            identifiers,
            expressions,
            statements,
            root,
        }
    }

    pub fn statements(&self) -> &Vec<StmtId> {
        &self.root
    }

    pub fn identifier(&self, id: IdentId) -> &String {
        &self.identifiers[id.id()]
    }

    pub fn expression(&self, id: ExprId) -> &Expression {
        &self.expressions[id.id()]
    }

    pub fn statement(&self, id: StmtId) -> &Statement {
        &self.statements[id.id()]
    }
}

pub struct PrettyPrint<'a, T> {
    indent: usize,
    ast: &'a Ast,
    id: T,
}

impl<'a, T> PrettyPrint<'a, T> {
    pub fn new(ast: &'a Ast, id: T) -> Self {
        Self { indent: 0, ast, id }
    }

    fn new_with_ident(ast: &'a Ast, id: T, ident: usize) -> Self {
        Self {
            indent: ident,
            ast,
            id,
        }
    }
}

impl Display for PrettyPrint<'_, StmtId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        match self.ast.statement(self.id).kind() {
            StatementKind::Expression(expr) => {
                let semi = match self.ast.expression(*expr).kind() {
                    ExpressionKind::If(_, _, _) | ExpressionKind::While(_, _)=> "",
                    _ => ";",
                };
                writeln!(
                    f,
                    "{indent}{}{semi}",
                    PrettyPrint::new_with_ident(self.ast, *expr, self.indent)
                )
            }
            StatementKind::Let(ident, expr) => {
                writeln!(
                    f,
                    "{indent}let {} = {};",
                    PrettyPrint::new(self.ast, ident.id()),
                    PrettyPrint::new(self.ast, *expr),
                )
            }
            StatementKind::Ret(expr) => {
                writeln!(f, "{indent}ret {};", PrettyPrint::new(self.ast, *expr))
            }
            StatementKind::LetFn(ident, params, stmts) => {
                write!(f, "{indent}let {}(", PrettyPrint::new(self.ast, ident.id()))?;
                for (i, param) in params.iter().enumerate() {
                    write!(f, "{}", PrettyPrint::new(self.ast, param.id()))?;
                    if i < params.len() - 1 {
                        write!(f, ", ")?;
                    }
                }
                writeln!(f, ") = {{")?;
                for stmt in stmts {
                    write!(
                        f,
                        "{indent}{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                writeln!(f, "{indent}}}")
            }
        }
    }
}

impl Display for PrettyPrint<'_, ExprId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        match self.ast.expression(self.id).kind() {
            ExpressionKind::Assignment(ident, expr) => {
                write!(
                    f,
                    "{} = {}",
                    PrettyPrint::new(self.ast, ident.id()),
                    PrettyPrint::new(self.ast, *expr)
                )
            }
            ExpressionKind::If(cond, true_branch, false_branch) => {
                writeln!(f, "if {} {{", PrettyPrint::new(self.ast, *cond))?;
                for stmt in true_branch {
                    write!(
                        f,
                        "{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                if !false_branch.is_empty() {
                    writeln!(f, "}}")?;
                    writeln!(f, "else {{")?;
                    for stmt in false_branch {
                        write!(
                            f,
                            "{}",
                            PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                        )?;
                    }
                }
                write!(f, "{indent}}}")
            }
            ExpressionKind::Literal(lit) => match lit.kind() {
                LiteralKind::Boolean(b) => write!(f, "{b}"),
                LiteralKind::Identifier(ident) => {
                    write!(f, "{}", PrettyPrint::new(self.ast, ident.id()))
                }
                LiteralKind::Number(n) => write!(f, "{n}"),
            },
            ExpressionKind::Binary(left, op, right) => {
                write!(
                    f,
                    "{} {} {}",
                    PrettyPrint::new(self.ast, *left),
                    op.kind(),
                    PrettyPrint::new(self.ast, *right)
                )
            }
            ExpressionKind::FunctionCall(ident, params) => {
                write!(f, "{}(", PrettyPrint::new(self.ast, ident.id()))?;
                for (i, expr) in params.iter().enumerate() {
                    write!(f, "{}", PrettyPrint::new(self.ast, *expr))?;
                    if i < params.len() - 1 {
                        write!(f, ", ")?;
                    }
                }
                write!(f, ")")
            }
            ExpressionKind::Unary(op, expr) => {
                write!(f, "{} {}", op.kind(), PrettyPrint::new(self.ast, *expr))
            }
            ExpressionKind::While(cond, stmts) => {
                writeln!(f, "while {} {{", PrettyPrint::new(self.ast, *cond))?;
                for stmt in stmts {
                    write!(
                        f,
                        "{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                writeln!(f, "}}")
            }
        }
    }
}

impl Display for PrettyPrint<'_, IdentId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        write!(f, "{indent}{}", self.ast.identifier(self.id))
    }
}
