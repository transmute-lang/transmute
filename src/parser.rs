use crate::ast::expression::{Expression, ExpressionKind};
use crate::ast::identifier::Identifier;
use crate::ast::literal::{Literal, LiteralKind};
use crate::ast::operators::{BinaryOperator, BinaryOperatorKind, UnaryOperator, UnaryOperatorKind};
use crate::ast::statement::{Statement, StatementKind};
use crate::ast::Ast;
use crate::error::Error;
use crate::lexer::{Lexer, PeekableLexer, Span, TokenKind};

pub struct Parser<'a> {
    lexer: PeekableLexer<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: Lexer<'a>) -> Self {
        Self {
            lexer: PeekableLexer::new(lexer, 1),
        }
    }

    pub fn parse(&mut self) -> Result<Ast, Error> {
        let mut statements = Vec::new();
        while let Ok(token) = self.lexer.peek() {
            if token.kind() == &TokenKind::Eof {
                break;
            }

            statements.push(self.parse_statement()?);
        }

        Ok(Ast::new(statements))
    }

    /// Parses the following:
    /// ```
    /// 'let ident = expr ;
    /// 'let ident ( expr , ... ) = expr ;
    /// 'let ident ( expr , ... ) = { expr ; ... }
    /// 'expr ;
    /// 'ret expr ;
    /// ```
    fn parse_statement(&mut self) -> Result<Statement, Error> {
        // fixme: duplicated several times
        let token = match self.lexer.peek() {
            Ok(token) => token,
            Err(_) => Err(self.lexer.next().unwrap_err())?,
        };

        match token.kind() {
            TokenKind::Let => {
                let let_token = self.lexer.next().expect("we just peeked it");
                let identifier_token = self.lexer.next()?;
                let identifier = match identifier_token.kind() {
                    TokenKind::Identifier(ident) => {
                        Identifier::new(ident.to_string(), identifier_token.span().clone())
                    }
                    _ => todo!(
                        "parse_statement: error handling {:?} at {:?}",
                        identifier_token.kind(),
                        identifier_token.span()
                    ),
                };

                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };

                if token.kind() == &TokenKind::OpenParenthesis {
                    // let ident '( expr , ... ) = expr ;
                    return self.parse_function(let_token.span(), identifier);
                }

                // fixme: pattern duplicated several times
                let equal_token = self.lexer.next()?;
                if equal_token.kind() != &TokenKind::Equal {
                    todo!(
                        "parse_statement: error handling {:?}; expected = at {:?}",
                        equal_token.kind(),
                        equal_token.span()
                    )
                }

                let expression = self.parse_expression()?;

                // fixme: duplicated several times
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }

                let span = let_token.span().extend_to(semicolon_token.span());
                Ok(Statement::new(
                    StatementKind::Let(identifier, expression),
                    span,
                ))
            }
            TokenKind::Ret => {
                let ret_token = self.lexer.next().expect("we just peeked it");
                let expression = self.parse_expression()?;
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }

                Ok(Statement::new(
                    StatementKind::Ret(expression),
                    ret_token.span().extend_to(semicolon_token.span()),
                ))
            }
            TokenKind::Minus
            | TokenKind::Number(_)
            | TokenKind::OpenParenthesis
            | TokenKind::Identifier(_) => {
                let expression = self.parse_expression()?;
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }
                let span = expression.span().extend_to(semicolon_token.span());
                Ok(Statement::new(StatementKind::Expression(expression), span))
            }
            _ => todo!(
                "parse_statement: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        }
    }

    /// Parses the following:
    /// ```
    /// let ident '( expr , ... ) = expr ;
    /// let ident '( expr , ... ) = { expr ; ... }
    /// ```
    fn parse_function(&mut self, span: &Span, identifier: Identifier) -> Result<Statement, Error> {
        // let name '( expr , ... ) = expr ;
        let open_parenthesis_token = self.lexer.next().expect("present, as we peeked it already");
        assert_eq!(open_parenthesis_token.kind(), &TokenKind::OpenParenthesis);

        let mut parameters = Vec::new();
        let mut comma_seen = true;

        loop {
            let token = self.lexer.next()?;
            match token.kind() {
                TokenKind::CloseParenthesis => {
                    break;
                }
                TokenKind::Comma if !comma_seen => {
                    comma_seen = true;
                }
                TokenKind::Identifier(ident) if comma_seen => {
                    parameters.push(Identifier::new(ident.to_string(), token.span().clone()));
                    comma_seen = false;
                }
                _ => todo!(
                    "parse_function: error handling {:?} at {:?}",
                    token.kind(),
                    token.span()
                ),
            }
        }

        // let name ( expr , ... ) '= expr ;
        let token = self.lexer.next()?;
        if token.kind() != &TokenKind::Equal {
            todo!(
                "parse_function: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            )
        }

        let token = match self.lexer.peek() {
            Ok(token) => token,
            Err(_) => Err(self.lexer.next().unwrap_err())?,
        };

        let (statements, end_span) = if token.kind() == &TokenKind::OpenCurlyBracket {
            // let name ( expr , ... ) = '{ expr ; ... }
            let _open_curly_bracket_token = self.lexer.next().expect("we just peeked it");

            let mut statements = Vec::new();

            let span = loop {
                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };
                match token.kind() {
                    TokenKind::CloseCurlyBracket => {
                        let token = self.lexer.next().expect("we just peeked it");
                        break token.span().clone();
                    }
                    _ => {
                        statements.push(self.parse_statement()?);
                    }
                }
            };

            (statements, span)
        } else {
            // let name ( expr , ... ) = 'expr ;
            let expression = self.parse_expression()?;

            let semicolon_token = self.lexer.next()?;
            if semicolon_token.kind() != &TokenKind::Semicolon {
                todo!(
                    "parse_statement: error handling {:?}; expected = at {:?}",
                    semicolon_token.kind(),
                    semicolon_token.span()
                )
            }

            let end_span = expression.span().extend_to(semicolon_token.span());

            (
                vec![Statement::new(
                    StatementKind::Expression(expression),
                    end_span.clone(),
                )],
                end_span,
            )
        };

        Ok(Statement::new(
            StatementKind::LetFn(identifier, parameters, statements),
            span.extend_to(&end_span),
        ))
    }

    /// Parses the following:
    /// ```
    /// 'expr
    /// ```
    fn parse_expression(&mut self) -> Result<Expression, Error> {
        self.parse_expression_with_precedence(Precedence::Lowest)
    }

    fn parse_expression_with_precedence(
        &mut self,
        precedence: Precedence,
    ) -> Result<Expression, Error> {
        let token = self.lexer.next()?;

        let mut expression = match token.kind() {
            TokenKind::Identifier(ident) => {
                let identifier = Identifier::new(ident.clone(), token.span().clone());

                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };

                match token.kind() {
                    &TokenKind::OpenParenthesis => self.parse_function_call(identifier)?,
                    _ => {
                        let span = identifier.span().clone();
                        Expression::from(Literal::new(LiteralKind::Identifier(identifier), span))
                    }
                }
            }
            TokenKind::Number(n) => {
                Expression::from(Literal::new(LiteralKind::Number(*n), token.span().clone()))
            }
            TokenKind::Minus => {
                let expression = self.parse_expression_with_precedence(Precedence::Prefix)?;
                let span = token.span().extend_to(expression.span());
                Expression::new(
                    ExpressionKind::Unary(
                        UnaryOperator::new(UnaryOperatorKind::Minus, token.span().clone()),
                        Box::new(expression),
                    ),
                    span,
                )
            }
            TokenKind::OpenParenthesis => {
                let open_loc = token.span();

                let expression = self.parse_expression()?;
                let token = self.lexer.next()?;
                if token.kind() == &TokenKind::CloseParenthesis {
                    expression
                } else {
                    todo!(
                        "parse_expression_with_precedence: error handling {:?}; expected ) at {:?} for parenthesis open at {:?}",
                        token.kind(),
                        token.span(),
                        open_loc
                    )
                }
            }
            _ => todo!(
                "parse_expression_with_precedence: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        };

        while let Some(next_precedence) = self
            .lexer
            .peek()
            .as_ref()
            .map(|t| Into::<Option<Precedence>>::into(t.kind()))
            .unwrap_or_default()
        {
            if precedence < next_precedence {
                expression = self.parse_infix_expression(expression, next_precedence)?;
            } else {
                break;
            }
        }

        Ok(expression)
    }

    /// Parses the following:
    /// ```
    /// identifier ( expr , ... )
    /// ```
    fn parse_function_call(&mut self, identifier: Identifier) -> Result<Expression, Error> {
        // identifier '( expr , ... )
        let open_parenthesis_token = self.lexer.next()?;
        assert_eq!(open_parenthesis_token.kind(), &TokenKind::OpenParenthesis);

        let mut arguments = Vec::new();
        let mut comma_seen = true;

        let span = loop {
            let token = match self.lexer.peek() {
                Ok(token) => token,
                Err(_) => Err(self.lexer.next().unwrap_err())?,
            };
            match *token.kind() {
                TokenKind::CloseParenthesis => {
                    let token = self.lexer.next().expect("we just peeked it");
                    break identifier.span().extend_to(token.span());
                }
                TokenKind::Comma if !comma_seen => {
                    self.lexer.next().expect("we just peeked it");
                    comma_seen = true;
                }
                TokenKind::Comma if comma_seen => {
                    todo!(
                        "parse_function_call: error handling {:?} at {:?}",
                        token.kind(),
                        token.span()
                    )
                }
                _ if comma_seen => {
                    arguments.push(self.parse_expression()?);
                    comma_seen = false;
                }
                _ => todo!(
                    "parse_function_call: error handling {:?} at {:?}",
                    token.kind(),
                    token.span()
                ),
            }
        };

        // identifier ( expr , ... ) '

        Ok(Expression::new(
            ExpressionKind::MethodCall(identifier, arguments),
            span,
        ))
    }

    fn parse_infix_expression(
        &mut self,
        left: Expression,
        precedence: Precedence,
    ) -> Result<Expression, Error> {
        let operator = self
            .parse_operator()
            .expect("there must be an operator, we got a precedence");
        let right = self.parse_expression_with_precedence(precedence)?;

        let span = left.span().extend_to(right.span());

        Ok(Expression::new(
            ExpressionKind::Binary(Box::new(left), operator, Box::new(right)),
            span,
        ))
    }

    fn parse_operator(&mut self) -> Result<BinaryOperator, Error> {
        let token = self.lexer.next()?;

        match token.kind() {
            TokenKind::Plus => Ok(BinaryOperator::new(
                BinaryOperatorKind::Addition,
                token.span().clone(),
            )),
            TokenKind::Minus => Ok(BinaryOperator::new(
                BinaryOperatorKind::Subtraction,
                token.span().clone(),
            )),
            TokenKind::Star => Ok(BinaryOperator::new(
                BinaryOperatorKind::Multiplication,
                token.span().clone(),
            )),
            TokenKind::Slash => Ok(BinaryOperator::new(
                BinaryOperatorKind::Division,
                token.span().clone(),
            )),
            _ => todo!(
                "parse_operator: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        }
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq)]
enum Precedence {
    Lowest,
    Sum,
    Product,
    Prefix,
}

impl From<&TokenKind> for Option<Precedence> {
    fn from(kind: &TokenKind) -> Self {
        match kind {
            TokenKind::Plus => Some(Precedence::Sum),
            TokenKind::Minus => Some(Precedence::Sum),
            TokenKind::Star => Some(Precedence::Product),
            TokenKind::Slash => Some(Precedence::Product),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ast::Ast;
    use crate::lexer::Span;

    #[test]
    fn expression_literal_number() {
        let lexer = Lexer::new("42");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Literal(Literal::new(LiteralKind::Number(42), Span::new(1, 1, 0, 2))),
            Span::new(1, 1, 0, 2),
        );

        assert_eq!(expected, actual);
    }

    #[test]
    fn expression_literal_identifier() {
        let mut parser = Parser::new(Lexer::new("forty_two"));

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Literal(Literal::new(
                LiteralKind::Identifier(Identifier::new(
                    "forty_two".to_string(),
                    Span::new(1, 1, 0, 9),
                )),
                Span::new(1, 1, 0, 9),
            )),
            Span::new(1, 1, 0, 9),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn expression() {
        let lexer = Lexer::new("2 + 20 * 2");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(2),
                    Span::new(1, 1, 0, 1),
                ))),
                BinaryOperator::new(BinaryOperatorKind::Addition, Span::new(1, 3, 2, 1)),
                Box::new(Expression::new(
                    ExpressionKind::Binary(
                        Box::new(Expression::from(Literal::new(
                            LiteralKind::Number(20),
                            Span::new(1, 5, 4, 2),
                        ))),
                        BinaryOperator::new(
                            BinaryOperatorKind::Multiplication,
                            Span::new(1, 8, 7, 1),
                        ),
                        Box::new(Expression::from(Literal::new(
                            LiteralKind::Number(2),
                            Span::new(1, 10, 9, 1),
                        ))),
                    ),
                    Span::new(1, 5, 4, 6),
                )),
            ),
            Span::new(1, 1, 0, 10),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn binary_operator_minus() {
        let lexer = Lexer::new("43 - 1");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(43),
                    Span::new(1, 1, 0, 2),
                ))),
                BinaryOperator::new(BinaryOperatorKind::Subtraction, Span::new(1, 4, 3, 1)),
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(1),
                    Span::new(1, 6, 5, 1),
                ))),
            ),
            Span::new(1, 1, 0, 6),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn prefix_minus() {
        let mut parser = Parser::new(Lexer::new("- 40 * 2"));

        let actual = parser.parse_expression().expect("expression is valid");
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::new(
                    ExpressionKind::Unary(
                        UnaryOperator::new(UnaryOperatorKind::Minus, Span::new(1, 1, 0, 1)),
                        Box::new(Expression::new(
                            ExpressionKind::Literal(Literal::new(
                                LiteralKind::Number(40),
                                Span::new(1, 3, 2, 2),
                            )),
                            Span::new(1, 3, 2, 2),
                        )),
                    ),
                    Span::new(1, 1, 0, 4),
                )),
                BinaryOperator::new(BinaryOperatorKind::Multiplication, Span::new(1, 6, 5, 1)),
                Box::new(Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(2),
                        Span::new(1, 8, 7, 1),
                    )),
                    Span::new(1, 8, 7, 1),
                )),
            ),
            Span::new(1, 1, 0, 8),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn root() {
        let lexer = Lexer::new("40 + 2;");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse().unwrap();
        let expected = Ast::new(vec![Statement::new(
            StatementKind::Expression(Expression::new(
                ExpressionKind::Binary(
                    Box::new(Expression::from(Literal::new(
                        LiteralKind::Number(40),
                        Span::new(1, 1, 0, 2),
                    ))),
                    BinaryOperator::new(BinaryOperatorKind::Addition, Span::new(1, 4, 3, 1)),
                    Box::new(Expression::from(Literal::new(
                        LiteralKind::Number(2),
                        Span::new(1, 6, 5, 1),
                    ))),
                ),
                Span::new(1, 1, 0, 6),
            )),
            Span::new(1, 1, 0, 7),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    #[should_panic]
    fn missing_right_parenthesis() {
        let mut parser = Parser::new(Lexer::new("(42"));
        let _ = parser.parse();
    }

    #[test]
    fn let_statement() {
        let mut parser = Parser::new(Lexer::new("let forty_two = 42;"));

        let actual = parser.parse_statement().expect("statement is valid");
        let expected = Statement::new(
            StatementKind::Let(
                Identifier::new("forty_two".to_string(), Span::new(1, 5, 4, 9)),
                Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(42),
                        Span::new(1, 17, 16, 2),
                    )),
                    Span::new(1, 17, 16, 2),
                ),
            ),
            Span::new(1, 1, 0, 19),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn two_statements() {
        let mut parser = Parser::new(Lexer::new("let forty_two = 42; forty_two;"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![
            Statement::new(
                StatementKind::Let(
                    Identifier::new("forty_two".to_string(), Span::new(1, 5, 4, 9)),
                    Expression::new(
                        ExpressionKind::Literal(Literal::new(
                            LiteralKind::Number(42),
                            Span::new(1, 17, 16, 2),
                        )),
                        Span::new(1, 17, 16, 2),
                    ),
                ),
                Span::new(1, 1, 0, 19),
            ),
            Statement::new(
                StatementKind::Expression(Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Identifier(Identifier::new(
                            "forty_two".to_string(),
                            Span::new(1, 21, 20, 9),
                        )),
                        Span::new(1, 21, 20, 9),
                    )),
                    Span::new(1, 21, 20, 9),
                )),
                Span::new(1, 21, 20, 10),
            ),
        ]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_statement() {
        let mut parser = Parser::new(Lexer::new("let times_two(a) = a * 2;"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![Statement::new(
            StatementKind::LetFn(
                Identifier::new("times_two".to_string(), Span::new(1, 5, 4, 9)),
                vec![Identifier::new("a".to_string(), Span::new(1, 15, 14, 1))],
                vec![Statement::new(
                    StatementKind::Expression(Expression::new(
                        ExpressionKind::Binary(
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Identifier(Identifier::new(
                                        "a".to_string(),
                                        Span::new(1, 20, 19, 1),
                                    )),
                                    Span::new(1, 20, 19, 1),
                                )),
                                Span::new(1, 20, 19, 1),
                            )),
                            BinaryOperator::new(
                                BinaryOperatorKind::Multiplication,
                                Span::new(1, 22, 21, 1),
                            ),
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Number(2),
                                    Span::new(1, 24, 23, 1),
                                )),
                                Span::new(1, 24, 23, 1),
                            )),
                        ),
                        Span::new(1, 20, 19, 5),
                    )),
                    Span::new(1, 20, 19, 6),
                )],
            ),
            Span::new(1, 1, 0, 25),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_statements() {
        let mut parser = Parser::new(Lexer::new("let times_two(a) = { a * 2; }"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![Statement::new(
            StatementKind::LetFn(
                Identifier::new("times_two".to_string(), Span::new(1, 5, 4, 9)),
                vec![Identifier::new("a".to_string(), Span::new(1, 15, 14, 1))],
                vec![Statement::new(
                    StatementKind::Expression(Expression::new(
                        ExpressionKind::Binary(
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Identifier(Identifier::new(
                                        "a".to_string(),
                                        Span::new(1, 22, 21, 1),
                                    )),
                                    Span::new(1, 22, 21, 1),
                                )),
                                Span::new(1, 22, 21, 1),
                            )),
                            BinaryOperator::new(
                                BinaryOperatorKind::Multiplication,
                                Span::new(1, 24, 23, 1),
                            ),
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Number(2),
                                    Span::new(1, 26, 25, 1),
                                )),
                                Span::new(1, 26, 25, 1),
                            )),
                        ),
                        Span::new(1, 22, 21, 5),
                    )),
                    Span::new(1, 22, 21, 6),
                )],
            ),
            Span::new(1, 1, 0, 29),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_call() {
        let mut parser = Parser::new(Lexer::new("times_two(21)"));
        let actual = parser.parse_expression().expect("expression is valid");
        let expected = Expression::new(
            ExpressionKind::MethodCall(
                Identifier::new("times_two".to_string(), Span::new(1, 1, 0, 9)),
                vec![Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(21),
                        Span::new(1, 11, 10, 2),
                    )),
                    Span::new(1, 11, 10, 2),
                )],
            ),
            Span::new(1, 1, 0, 13),
        );
        assert_eq!(actual, expected);
    }

    #[test]
    fn ret() {
        let mut parser = Parser::new(Lexer::new("ret 42;"));
        let actual = parser.parse_statement().expect("statement is valid");
        let expected = Statement::new(
            StatementKind::Ret(Expression::new(
                ExpressionKind::Literal(Literal::new(
                    LiteralKind::Number(42),
                    Span::new(1, 5, 4, 2),
                )),
                Span::new(1, 5, 4, 2),
            )),
            Span::new(1, 1, 0, 7),
        );
        assert_eq!(actual, expected);
    }
}
