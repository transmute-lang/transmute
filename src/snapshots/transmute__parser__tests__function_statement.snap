---
source: src/parser.rs
expression: "Parser::new(Lexer::new(\"let times_two(a: number) = a * 2;\")).parse()"
---
(
    Ast {
        identifiers: [
            "times_two",
            "a",
            "number",
        ],
        identifier_refs: [
            IdentifierRef {
                id: IdentRefId {
                    id: 0,
                },
                ident: Identifier {
                    id: IdentId {
                        id: 1,
                    },
                    span: 1:28; [27, 28],
                },
                symbol: UnresolvedSymbol,
            },
        ],
        expressions: [
            Expression {
                id: ExprId {
                    id: 0,
                },
                kind: Literal(
                    Literal {
                        kind: Identifier(
                            IdentRefId {
                                id: 0,
                            },
                        ),
                        span: 1:28; [27, 28],
                    },
                ),
                span: 1:28; [27, 28],
            },
            Expression {
                id: ExprId {
                    id: 1,
                },
                kind: Literal(
                    Literal {
                        kind: Number(
                            2,
                        ),
                        span: 1:32; [31, 32],
                    },
                ),
                span: 1:32; [31, 32],
            },
            Expression {
                id: ExprId {
                    id: 2,
                },
                kind: Binary(
                    ExprId {
                        id: 0,
                    },
                    BinaryOperator {
                        kind: Multiplication,
                        span: 1:30; [29, 30],
                    },
                    ExprId {
                        id: 1,
                    },
                ),
                span: 1:28; [27, 32],
            },
            Expression {
                id: ExprId {
                    id: 3,
                },
                kind: Block(
                    [
                        StmtId {
                            id: 0,
                        },
                    ],
                ),
                span: 1:28; [27, 33],
            },
        ],
        statements: [
            Statement {
                id: StmtId {
                    id: 0,
                },
                kind: Expression(
                    ExprId {
                        id: 2,
                    },
                ),
                span: 1:28; [27, 33],
            },
            Statement {
                id: StmtId {
                    id: 1,
                },
                kind: LetFn(
                    Identifier {
                        id: IdentId {
                            id: 0,
                        },
                        span: 1:5; [4, 13],
                    },
                    [
                        Parameter {
                            identifier: Identifier {
                                id: IdentId {
                                    id: 1,
                                },
                                span: 1:15; [14, 15],
                            },
                            ty: Identifier {
                                id: IdentId {
                                    id: 2,
                                },
                                span: 1:18; [17, 23],
                            },
                            span: 1:15; [14, 23],
                        },
                    ],
                    None,
                    ExprId {
                        id: 3,
                    },
                ),
                span: 1:1; [0, 33],
            },
        ],
        root: [
            StmtId {
                id: 1,
            },
        ],
        implicit_ret: PhantomData<transmute::ast::WithImplicitRet>,
    },
    Diagnostics {
        diagnostics: [],
    },
)
